app/store.go

```diff
@@ -0,0 +1,186 @@
+package app
+
+import (
+	"bytes"
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	abci "github.com/tendermint/abci/types"
+	"github.com/tendermint/go-wire"
+	"github.com/tendermint/merkleeyes/iavl"
+	cmn "github.com/tendermint/tmlibs/common"
+	dbm "github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/log"
+
+	"github.com/tendermint/basecoin/state"
+)
+
+// Store contains the merkle tree, and all info to handle abci requests
+type Store struct {
+	state.State
+	height    uint64
+	hash      []byte
+	persisted bool
+
+	logger log.Logger
+}
+
+var stateKey = []byte("merkle:state") // Database key for merkle tree save value db values
+
+// ChainState contains the latest Merkle root hash and the number of times `Commit` has been called
+type ChainState struct {
+	Hash   []byte
+	Height uint64
+}
+
+// NewStore initializes an in-memory IAVLTree, or attempts to load a persistant
+// tree from disk
+func NewStore(dbName string, cacheSize int, logger log.Logger) *Store {
+	// start at 1 so the height returned by query is for the
+	// next block, ie. the one that includes the AppHash for our current state
+	initialHeight := uint64(1)
+
+	// Non-persistent case
+	if dbName == "" {
+		tree := iavl.NewIAVLTree(
+			0,
+			nil,
+		)
+		return &Store{
+			State:  state.NewState(tree, false),
+			height: initialHeight,
+			logger: logger,
+		}
+	}
+
+	// Expand the path fully
+	dbPath, err := filepath.Abs(dbName)
+	if err != nil {
+		panic(fmt.Sprintf("Invalid Database Name: %s", dbName))
+	}
+
+	// Some external calls accidently add a ".db", which is now removed
+	dbPath = strings.TrimSuffix(dbPath, path.Ext(dbPath))
+
+	// Split the database name into it's components (dir, name)
+	dir := path.Dir(dbPath)
+	name := path.Base(dbPath)
+
+	// Make sure the path exists
+	empty, _ := cmn.IsDirEmpty(dbPath + ".db")
+
+	// Open database called "dir/name.db", if it doesn't exist it will be created
+	db := dbm.NewDB(name, dbm.LevelDBBackendStr, dir)
+	tree := iavl.NewIAVLTree(cacheSize, db)
+
+	var chainState ChainState
+	if empty {
+		logger.Info("no existing db, creating new db")
+		chainState = ChainState{
+			Hash:   tree.Save(),
+			Height: initialHeight,
+		}
+		db.Set(stateKey, wire.BinaryBytes(chainState))
+	} else {
+		logger.Info("loading existing db")
+		eyesStateBytes := db.Get(stateKey)
+		err = wire.ReadBinaryBytes(eyesStateBytes, &chainState)
+		if err != nil {
+			logger.Error("error reading MerkleEyesState", "err", err)
+			panic(err)
+		}
+		tree.Load(chainState.Hash)
+	}
+
+	return &Store{
+		State:     state.NewState(tree, true),
+		height:    chainState.Height,
+		hash:      chainState.Hash,
+		persisted: true,
+		logger:    logger,
+	}
+}
+
+// CloseDB closes the database
+// func (s *Store) CloseDB() {
```

Comment:
It was never really called anywhere, so I guess it can be removed. I'll delete it fully.